// @authentic/mwc-card/mwc-card.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
/**
@license
Copyright 2018 Google Inc. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { LitElement, customElement, html, property, classMap, findAssignedElements, query } from "@authentic/mwc-base/base-element";
import { ripple } from "@authentic/mwc-ripple/ripple-directive";
import { style } from "./mwc-card-css";
import "@authentic/mwc-icon/mwc-icon-font";
let Card = class Card extends LitElement {
    constructor() {
        super(...arguments);
        /**
         * Optional. Default value is false. Removes the shadow and displays a hairline outline instead.
         */
        this.outlined = false;
        /**
         * Optional. Define the aspect ratio for the main image, it can be 'square' or '16-9'
         */
        this.aspectRatio = '';
    }
    get headerEls() {
        return this.slotHeaderEl && findAssignedElements(this.slotHeaderEl, '*');
    }
    get mediaEls() {
        return this.slotMediaEl && findAssignedElements(this.slotMediaEl, '*');
    }
    get contentEls() {
        return this.slotContentEl && findAssignedElements(this.slotContentEl, '*');
    }
    get actionButtons() {
        return this.slotActionButtonsEl && findAssignedElements(this.slotActionButtonsEl, 'mwc-button');
    }
    get actionIcons() {
        return this.slotActionIconsEl && findAssignedElements(this.slotActionIconsEl, 'mwc-icon-button');
    }
    /**
     * Used to render the lit-html TemplateResult to the element's DOM
     */
    render() {
        const styles = {
            'mdc-card': true,
            'mdc-card--outlined': this.outlined
        };
        const mediaStyles = {
            'mdc-card__media': true,
            [`mdc-card__media--${this.aspectRatio}`]: this.aspectRatio !== '',
        };
        return html `
      <div class="${classMap(styles)}">
        <div class="mdc-card__primary-action" tabindex="0" .ripple="${ripple({ unbounded: false })}">
          <slot name="header"></slot>
          
          <div class="${classMap(mediaStyles)}">
            <div class="mdc-card__media-content">
              <slot name="media"></slot>
            </div>
          </div>

          <slot name="content"></slot>
        </div>

        <slot></slot>

        <div class="mdc-card__actions">
          <div class="mdc-card__action-buttons">
            <slot name="action-buttons"></slot>
          </div>

          <div class="mdc-card__action-icons">
            <slot name="action-icons"></slot>
          </div>
        </div>
      </div>
    `;
    }
    /**
     * This method is invoked whenever the card is updated
     * @param _changedProperties Map of changed properties with old values
     */
    updated(_changedProperties) {
        super.updated(_changedProperties);
        const _hasHeaderEls = this.headerEls.length > 0;
        const _hasMediaEls = this.mediaEls.length > 0;
        const _hasContentEls = this.contentEls.length > 0;
        const _hasPrimaryAction = _hasHeaderEls || _hasMediaEls || _hasContentEls;
        const _hasActionButtons = this.actionButtons.length > 0;
        const _hasActionIcons = this.actionIcons.length > 0;
        const _hasActions = _hasActionButtons || _hasActionIcons;
        this.primaryActionEl.classList.toggle('mdc-card__primary-action--empty', !_hasPrimaryAction);
        this.actionButtonsEl.classList.toggle('mdc-card__action-buttons--empty', !_hasActionButtons);
        this.actionIconsEl.classList.toggle('mdc-card__action-icons--empty', !_hasActionIcons);
        this.actionsEl.classList.toggle('mdc-card__actions--empty', !_hasActions);
    }
};
Card.styles = style;
__decorate([
    query('.mdc-card__primary-action')
], Card.prototype, "primaryActionEl", void 0);
__decorate([
    query('.mdc-card__actions')
], Card.prototype, "actionsEl", void 0);
__decorate([
    query('.mdc-card__action-buttons')
], Card.prototype, "actionButtonsEl", void 0);
__decorate([
    query('.mdc-card__action-icons')
], Card.prototype, "actionIconsEl", void 0);
__decorate([
    query('slot[name="header"]')
], Card.prototype, "slotHeaderEl", void 0);
__decorate([
    query('slot[name="media"]')
], Card.prototype, "slotMediaEl", void 0);
__decorate([
    query('slot[name="content"]')
], Card.prototype, "slotContentEl", void 0);
__decorate([
    query('slot[name="action-buttons"]')
], Card.prototype, "slotActionButtonsEl", void 0);
__decorate([
    query('slot[name="action-icons"]')
], Card.prototype, "slotActionIconsEl", void 0);
__decorate([
    property({ type: Boolean })
], Card.prototype, "outlined", void 0);
__decorate([
    property({ type: String })
], Card.prototype, "aspectRatio", void 0);
Card = __decorate([
    customElement("mwc-card")
], Card);
export { Card };
